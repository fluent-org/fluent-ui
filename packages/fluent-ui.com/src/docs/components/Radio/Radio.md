---
title: Radio
components: Radio
type: Inputs
langKey: "en"
---

# Radio

<p class="description">Radio buttons, let users select one option from two or more choices. Each option is represented by one Choice Group button; a user can select only one ChoiceGroup in a button group.</p>

## Controlled

```jsx
() => {
  const [checked, set] = React.useState('a')
  function handleChange(c) {
    set(c)
  }
  return (
    <>
      <Radio value='a' checked={checked === 'a'} onChange={handleChange} />
      <Radio value='b' checked={checked === 'b'} onChange={handleChange} />
      <Radio value='c' checked={checked === 'c'} onChange={handleChange} />
    </>
  )
}
```

## With Label

```jsx
() => {
  const [checked, set] = React.useState('a')
  function handleChange(c) {
    set(c)
  }
  return (
    <>
      <FormLabel label="a">
        <Radio value='a' checked={checked === 'a'} onChange={handleChange} />
      </FormLabel>
      <FormLabel label="b">
        <Radio value='b' checked={checked === 'b'} onChange={handleChange} />
      </FormLabel>
      <FormLabel label="b">
        <Radio value='c' checked={checked === 'c'} onChange={handleChange} />
      </FormLabel>
    </>
  )
}
```

## Disabled

```jsx
<>
  <Radio disabled />
  <Radio disabled checked />
</>
```

