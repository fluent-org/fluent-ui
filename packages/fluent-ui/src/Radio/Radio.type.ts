import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps, StyledProps } from '..'

export type RadioClassProps =
  | 'root'
  | 'checked'
  | 'disabled'
  | 'circle'
  | 'circleChecked'
  | 'circleDisabled'

export interface RadioProps
  extends StandardProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement, 'onChange'>,
    StyledProps {
  checked?: boolean
  value?: string
  onChange?: (checked: string) => void
  disabled?: boolean
}

export const RadioPropTypes = {
  checked: PropTypes.bool,
  value: PropTypes.string,
  onChange: PropTypes.func,
  disabled: PropTypes.bool
}
