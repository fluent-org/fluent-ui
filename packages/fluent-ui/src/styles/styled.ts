import styled from 'styled-components'
import { th, variant, system, breakpoints } from '@xstyled/system'

export { styled, system, th, variant, breakpoints }

export * from 'styled-components'
export * from '@xstyled/system'
// export * from '@xstyled/prop-types'

export * from './stylize'
