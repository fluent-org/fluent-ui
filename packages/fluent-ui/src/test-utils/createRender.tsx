import * as React from 'react'
import * as PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

interface WrapperProps {
  children: React.ReactNode
}

const Wrapper: React.FC<WrapperProps> = ({ children }): React.ReactElement => (
  <ThemeProvider theme={{}}>{children}</ThemeProvider>
)

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui: React.ReactElement, options: any) =>
  render(ui, { wrapper: Wrapper, ...options })

export { customRender as render }
