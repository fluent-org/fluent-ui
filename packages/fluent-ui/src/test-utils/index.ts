export * from '@testing-library/react'
// @ts-ignore
export * from './createRender'
