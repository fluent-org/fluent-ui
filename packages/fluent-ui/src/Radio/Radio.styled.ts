import { styled, css, th, variant } from '../styles/styled'

const checked = variant({
  prop: 'checked',
  default: false,
  variants: {
    true: css`
      color: ${th.color('white.default')};
      border-color: ${th.color('accent.default')};
      &:hover {
        border-color: ${th.color('accent.default')};
      }
    `,
    false: css`
      color: ${th.color('black.default')};
      border-color: ${th.color('primary.default')};
    `
  }
})

const disabled = variant({
  prop: 'disabled',
  default: false,
  variants: {
    true: css`
      border-color: ${th.color('primary.dark3')};
      cursor: not-allowed;
      pointer-events: none;
    `,
    false: css``
  }
})

export const StyledRadioWrapper = styled.div<{
  checked?: boolean
  disabled?: boolean
}>`
  width: 20px;
  height: 20px;
  border-radius: 50%;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid;
  transition: ${th.transition('radio')};
  &:hover {
    border-color: ${th.color('primary.dark3')};
  }

  ${checked}
  ${disabled}
`

export const StyledRadio = styled.input`
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: ${({ disabled }): string => (disabled ? 'not-allowed' : 'pointer')};
`

export const StyledCircle = styled.div<{
  disabled?: boolean
  checked?: boolean
}>`
  border-radius: 50%;
  transition: ${th.transition('radio')};
  background-color: ${({ disabled }): string =>
    disabled ? th.color('primary.dark3') : th.color('black.default')};
  ${variant({
    prop: 'checked',
    default: false,
    variants: {
      true: css`
        width: 12px;
        height: 12px;
        opacity: 1;
      `,
      false: css`
        width: 0;
        height: 0;
        opacity: 0;
      `
    }
  })}
`
