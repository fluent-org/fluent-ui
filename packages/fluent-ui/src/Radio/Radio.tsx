import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles } from '@fluent-ui/styles'
import { RadioClassProps, RadioProps, RadioPropTypes } from './Radio.type'
import { Theme } from '../styles'
import { styles } from './Radio.styled'

export const name = 'Radio'

const useStyles = createUseStyles<Theme, RadioClassProps>(styles, { name })

const Radio: React.FC<RadioProps> = React.forwardRef<HTMLInputElement, RadioProps>(
  (props, ref): React.ReactElement => {
    const {
      as: Component = 'div',
      className: classNameProp,
      checked,
      value,
      onChange,
      disabled,
      ...rest
    } = props
    const classes = useStyles(props)
    const className = classNames(
      classes.root,
      {
        [classes.checked]: checked,
        [classes.disabled]: disabled
      },
      classNameProp
    )
    const circleClassName = classNames(classes.circle, {
      [classes.circleChecked]: checked,
      [classes.circleDisable]: disabled
    })

    function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
      onChange && onChange(e.target.value)
    }
    return (
      <Component className={className}>
        <div className={circleClassName} />
        <input
          ref={ref}
          type="radio"
          checked={checked}
          value={value}
          onChange={handleChange}
          disabled={disabled}
          {...rest}
        />
      </Component>
    )
  }
)

Radio.displayName = `F${name}`

Radio.propTypes = RadioPropTypes

export default Radio
