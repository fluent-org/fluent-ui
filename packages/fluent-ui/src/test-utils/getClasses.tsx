import * as React from 'react'
import { createUseStyles } from '@fluent-ui/styles'
import { Styles } from 'jss'
import { render } from './createRender'

export function getClasses<T extends string>(styles: Styles, name: string): Record<T, string> {
  const useStyles = createUseStyles<T>(styles, { name })
  let classes: Record<T, string>
  function Util(props: any): React.ReactElement {
    classes = useStyles(props)
    return <div />
  }
  render(<Util />)
  // @ts-ignore
  return classes
}
