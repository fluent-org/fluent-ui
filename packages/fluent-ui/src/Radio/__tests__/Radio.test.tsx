import * as React from 'react'
import { fireEvent } from '@testing-library/react'
import { getClasses, render } from '../../test-utils'
import '@testing-library/jest-dom/extend-expect'

import Radio, { name } from '../Radio'
import { styles } from '../Radio.styled'
import { RadioClassProps } from '../Radio.type'

const classes = getClasses<RadioClassProps>(styles, name)

describe('Radio', (): void => {
  test('should be support ref', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { container } = render(<Radio ref={ref} />)
    const { current: element } = ref
    const RadioElement = container.querySelector(`input[type='Radio']`) as HTMLInputElement
    expect(element).toEqual(RadioElement)
  })

  test('should be controlled', async (): Promise<void> => {
    let checked = ''
    function handleChange(c: string): void {
      checked = c
    }
    const { getByTestId } = render(
      <>
        <Radio data-testid="a" value="a" checked={checked === 'a'} onChange={handleChange} />
        <Radio data-testid="b" value="b" checked={checked === 'b'} onChange={handleChange} />
        <Radio data-testid="c" value="c" checked={checked === 'c'} onChange={handleChange} />
      </>
    )
    fireEvent.click(getByTestId('a'))
    expect(checked).toEqual('a')
    fireEvent.click(getByTestId('b'))
    expect(checked).toEqual('b')
    fireEvent.click(getByTestId('c'))
    expect(checked).toEqual('c')
  })

  test('should be render a disabled Radio', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { container } = render(<Radio ref={ref} disabled={true} />)
    expect(container.firstChild).toHaveClass(classes.disabled)
    expect(ref.current!.disabled).toEqual(true)
  })
})
