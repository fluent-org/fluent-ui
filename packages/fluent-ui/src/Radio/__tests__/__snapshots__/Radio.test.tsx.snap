// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio basic 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  cursor: pointer;
}

.c4 {
  font-size: 0.8em;
  margin-left: 0;
}

.c1 {
  width: 20px;
  height: 20px;
  border-radius: 50%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid;
  -webkit-transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  color: #000000;
  border-color: #dbdbdb;
}

.c1:hover {
  border-color: #595959;
}

.c3 {
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: pointer;
}

.c2 {
  border-radius: 50%;
  -webkit-transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  background-color: #000000;
  width: 0;
  height: 0;
  opacity: 0;
}

<label
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <input
      className="c3"
      onChange={[Function]}
      type="radio"
    />
  </div>
  <span
    className="c4"
  />
</label>
`;

exports[`Radio prop disabled 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  cursor: not-allowed;
}

.c4 {
  font-size: 0.8em;
  margin-left: 0;
}

.c1 {
  width: 20px;
  height: 20px;
  border-radius: 50%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid;
  -webkit-transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  color: #000000;
  border-color: #dbdbdb;
  border-color: #595959;
  cursor: not-allowed;
  pointer-events: none;
}

.c1:hover {
  border-color: #595959;
}

.c3 {
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: not-allowed;
}

.c2 {
  border-radius: 50%;
  -webkit-transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  background-color: #595959;
  width: 0;
  height: 0;
  opacity: 0;
}

<label
  className="c0"
  disabled={true}
>
  <div
    className="c1"
    disabled={true}
  >
    <div
      className="c2"
      disabled={true}
    />
    <input
      className="c3"
      disabled={true}
      onChange={[Function]}
      type="radio"
    />
  </div>
  <span
    className="c4"
  />
</label>
`;
